File: Python_codes/binarytree.py

c = 0

class node():
    def __init__(self,data):
        self.data = data
        self.left = None
        self.right = None


def inorder(temp):
        if temp is None:
            return

        inorder(temp.left)
        print(temp.data)
        inorder(temp.right)

def insert(root, newnode):
    if root is None:
        root = newnode
    else:
        if root.data < newnode.data:
            if root.right is None:
                root.right = newnode
            else:
                insert(root.right, newnode)
        else:
            if root.left is None:
                root.left = newnode
            else:
                insert(root.left, newnode)


def search(root, key):
    if root is None:
        print("Not found")
        return

    if root.data == key:
        return root
    else:
        if root.data < key:
            return search(root.right,key)
        else:
            return search(root.left,key)

def findmin(root):
    if root is None:
        return

    while root.left is not None:
        root = root.left

    return root

def findmax(root):
    if root is None:
        return

    while root.right is not None:
        root = root.right

    return root

def kthlargestelement(root, k):
    global c
    if root is None:
        return

    #print(c)
    kthlargestelement(root.right, k)
    c = c+1
    if c == k:
        print(root.data)
        return
    
    kthlargestelement(root.left, k)

    

def deletenode(root,key):
    if root.data < key:
        deletenode(root.right, key)

    elif root.data > key:
        deletenode(root.left, key)

    else:
        if root.left is None and root.right is None:
            root = None
        elif root.left is None:
            temp = root.right
            root.right = None
            root.data = temp.data
        elif root.right is None:
            temp = root.left
            root.left = None
            root.data = temp.data
        else:
            current = root.left
            while current.left is not None:
                current = current.left

            root.data = current.data
            current = None
                
        



root = node(5)
e1 = node(2)
insert(root, e1)
e2 = node(3)
insert(root, e2)
e3 = node(7)
insert(root, e3)
e4 = node(8)
insert(root, e4)
inorder(root)

#inorder(search(root,2))
        
#deletenode(root, 5)
#inorder(root)
#min = findmin(root)
#print(min.data)
#max = findmax(root)
#print(max.data)
kthlargestelement(root, 2)

File: Tree/binary_tree.c
/******************************************************************************

                          Binary Tree

*******************************************************************************/

#include <stdio.h>
#include <stdlib.h>

struct node
{
    int data;
    struct node *left;
    struct node *right;
};

void print_preorder(struct node* root)
{
    if(root != NULL)
    {
        printf("\n %d", root->data);
        print_preorder(root->left);
        print_preorder(root->right);
        
    }
}

void deleteTree(struct node* root)
{
    if(root != NULL)
    {
        printf("\n %d", root->data);
        deleteTree(root->left);
        deleteTree(root->right);
        free(root);
        
    }
    
}

void insertNode(struct node **node, int data)
{
    if(*node == NULL)
    {
        struct node *temp = (struct node *) malloc(sizeof(struct node));
        temp->data = data;
        temp->left = NULL;
        temp->right = NULL;
        *node = temp;
        return;
    }
    
    if(data < (*node)->data)
    {
        insertNode(&(*node)->left, data);
    }
    else
    {
        insertNode(&(*node)->right, data);
    }
}

struct node* search(struct node* root, int data)
{
    if(root == NULL)
    {
        return NULL;
    }
    
    if(root->data == data)
    {
        return root;
    }
    
    if(data < root->data)
    {
        return search(root->left, data);
    }
    else
    {
        return search(root->right, data);
    }
    
    
}

int main()
{
    struct node *root = NULL;
    
    insertNode(&root, 9);
    insertNode(&root, 8);
    insertNode(&root, 7);
    insertNode(&root, 10);
    
    print_preorder(root);   
    
    struct node* temp = search(root, 5);
    
    if(temp)
    {
        printf("\n node found %d", temp->data);
    }
    else
    {
        printf("\n Not found");
    }
    
    deleteTree(root);

    return 0;
}
